ToneAlarm class refactoring to establish interface structure and tone_alarm library

mcsauder opened the pull request:
Hi,

This PR aims to make the ToneAlarm class implementations easier to maintain in the future by standardizing the existing 4 implementations of the ToneAlarm class.  Header files containing the class declarations and register settings have been broken out of each `/src/drivers/****/tone_alarm.cpp` file to simplify refactoring and breakout hardware specific implementation differences.

The ToneAlarmInterface namespace has been implemented for implementation specific methods as tone_alarm_interface libraries.

**Test data / coverage**
This PR has been bench tested on pixracrer (fmu-v4) and pixhawk 4 (fmu-v5).  Hardware testing of this PR is straightforward to validate as audible tones should indicate no differences between this PR andthe master branch firmware at bootup.

**NOTE** Hardware testing has not been validated for atmel same70 or nxp kinetis hardware.

**Describe problem solved by the proposed pull request**
ToneAlarm exists in 4 implementations currently.  This PR standardizes those implementations and creates an interface structure to avoid additional copy/paste implementations in the future.

Please let me know if you have any questions on this PR. Thanks!

-Mark

mcsauder commented:
***NOTE:*** This PR replaces the work started in #11132.

mcsauder commented:
Finally I've had a bit of success in implementing a tone_alarm library with a ToneAlarm base class for tone_alarm driver implementations to inherit from.

Bench testing on pixracer and pixhawk 4 succeeds as with previous; same70 and kinetis hardware testing still require validation.

Let me know if you have any questions on the revised implementation.

-Mark

LorenzMeier commented:
Awesome work! It's great to see the architecture improve at the core. I haven't done a detailed line-by-line review of the PR yet.

mcsauder commented:
Thanks Lorenz!  Hopefully I'll get past CI soon, I'm not sure why it's timing out. There is nothing abnormal on hardware; running top and free in the nsh show no differences between this PR and master on pixracer and pixhawk4 and they behave the same on the bench.

I tried accomplishing this structure with ModuleBase and having been through the exercise I think I will try to avoid ModuleBase as a generic driver base class in the future.

Let me know if you need anything from me on this PR!  Thanks again for your review.

mcsauder commented:
Hi @davids5 , could I impose on you, would you be willing to hardware test this PR on the kinetis and/or same70 for me?  CI tests were passing, I'll keep this PR rebased as master updates.

mcsauder commented:
@davids5 , certainly!

mcsauder commented:
@davids5 , I've made the changes you suggested during the dev call this morning.  If this looks good we can go with it, otherwise we can run with @dagar's branch, Either is fine by me!

davids5 commented:
@mcsauder looks good! Two things 1) Where is the IOCTL used? 2) Can we get to one main?

mcsauder commented:
Hi @davids5 , thanks!

1) I might not understand your question fully but the ToneAlaram* classes inherit from CDev, so `CDev::ioctl()` is inherited, although `CDev::ioctl` isn't used by these classes.  The register values look like they all get applied through `px4_configgpio()`, which invokes a board specific `configgpio()` method.

2) I'm glad you asked about getting down to a single main().... I've been trying all day but I haven't figured out a good solution to that yet. :}  I'll keep trying to get that done.

Thanks again for looking at this PR!

davids5 commented:
@mcsauder

> I might not understand your question fully but the ToneAlaram* classes inherit from CDev, so CDev::ioctl() is inherited, although CDev::ioctl isn't used by these classes. The register values look like they all get applied through px4_configgpio(), which invokes a board specific configgpio() method.

I think the  IOCTL was from the old world and only used in test. The https://github.com/PX4/Firmware/pull/11192/files#diff-c7afb89180cb59186904e4fa14038065R70 line 70-72  can come out and the the code in syscmnds test should use the uorb interface OR be removed.

2> I'm glad you asked about getting down to a single main().... I've been trying all day but I haven't figured out a good solution to that yet. :} I'll keep trying to get that done.

 have a look at https://github.com/PX4/Firmware/commit/8c117b682d2b911faa24d7b0f1c3f34740f4391b I think there is one main doen there.

mcsauder commented:
@davids5 , 
I don't think I can eliminate the #defines in drv_tone_alarm.h, they get used in tone_alarm.cpp, test_hrt.cpp, and mavlink_reciever.cpp, which I see now answers your question about IOCTL:

```
Firmware/src/lib/tone_alarm/tone_alarm.cpp:
   43  
   44  ToneAlarm::ToneAlarm() :
   45: 	CDev(TONE_ALARM0_DEVICE_PATH)
   46  {
   47  }

Firmware/src/modules/mavlink/mavlink_receiver.cpp:
 1697  
 1698  		if (*tune == 'M') {
 1699: 			int fd = px4_open(TONE_ALARM0_DEVICE_PATH, PX4_F_WRONLY);
 1700  
 1701  			if (fd >= 0) {

Firmware/src/systemcmds/tests/test_hrt.cpp:
  125  	unsigned long tone;
  126  
  127: 	fd = px4_open(TONE_ALARM0_DEVICE_PATH, O_WRONLY);
  128  
  129  	if (fd < 0) {
  130: 		printf("failed opening " TONE_ALARM0_DEVICE_PATH "\n");
  131  		goto out;
  132  	}
  140  	if (tone  == 0) {
  141: 		result = px4_ioctl(fd, TONE_SET_ALARM, TONE_STOP_TUNE);
  142  
  143  		if (result < 0) {
  ...
  150  
  151  	} else {
  152: 		result = px4_ioctl(fd, TONE_SET_ALARM, TONE_STOP_TUNE);
  153  
  154  		if (result < 0) {
  ...
  157  		}
  158  
  159: 		result = px4_ioctl(fd, TONE_SET_ALARM, tone);
  160  
  161  		if (result < 0) {
```
I've submitted a PR#11234 to break apart test_hrt.cpp into three files containing one test each to follow the convention in other test files, (rather than 3 tests in one file).  If that PR is accepted I can move `_TONE_ALARM_BASE` and `TONE_SET_ALARM` into `test_tone.cpp`, leaving only `TONEALARM0_DEVICE_PATH'.

In trying to get down to one main() function, I can create multiple implementations of the same interface class (as @dagar has done in his branch), the only drawback I see to that is that they all must share the same name; they cannot be uniquely identified.  Templating can't be used to pass a child class up to the parent because of `*gdev.  

Giving multiple implementations the same name works, it just feels a little hacky and I'm trying to figure out a way to avoid it.  Give me a day or two to keep thinking on it.

Thanks for your feedback!

I'm going to go ahead and follow @dagar's lead,

davids5 commented:
@mcsauder 

>I don't think I can eliminate the #defines in drv_tone_alarm.h, they get used in tone_alarm.cpp, test_hrt.cpp, and mavlink_reciever.cpp, which I see now answers your question about IOCTL:

I think this is broken code as it is today.  We want to not be a CDev, so the calling code has to be replaced with uOrb messages. So mavlink_receiver.cpp should look like what is done in src/systemcmds/tune_control/tune_control.cpp Line 220. The only caveat  is is a string pay is blocking as doen in tune_control.cpp, we may want to look at adding a buffer and a resource lock later. But delete the IOCTL callers and use uOrb for now.

On Firmware/src/systemcmds/tests/test_hrt.cpp. @dagar - OK to delete or do you want the uORB replacement?

>In trying to get down to one main() function, I can create multiple implementations of the same interface class (as @dagar has done in his branch), the only drawback I see to that is that they all must share the same name; they cannot be uniquely identified. Templating can't be used to pass a child class up to the parent because of *gdev. 

If I understand: That is 100% OK.  We should avoid Templates CRT because of the Bloat. 

We can have a call on Slack or Skype if you want more real time feedback. skype:david_s5

mcsauder commented:
Hi @davids5 , thanks for the help and offer to skype, I will take you up on that this week if my latest commit isn't there yet; take a look and give me your thoughts.

Although it is effective, the design pattern @dagar implemented somehow still feels a little gritty; it is compiling each lowest level driver implementation as a library and that seems sort of upside-down.  Because it also forces each hardware implementation to share the same name, I have a similar alternative that I hope is acceptable in my latest commit (pushed up).

I've placed all non-hardware specific method declarations and definitions in the `src/lib/drivers/ToneAlarm.h` header file. Remaining hardware specific methods are defined in their respective `src/drivers/***/tone_alarm/ToneAlarm.cpp` file.  This design pattern sets up `ToneAlarm.h` effectively as a header-only library and the header and source files carry the (same) name of the Class declared/defined within.

@dagar and @davids5, let me know if you still prefer @dagar's structure and I'll push things up that way. Otherwise, let me know your thoughts on this implementation!

Thanks!

mcsauder commented:
@dagar and @davids5 , I've rewritten and rebased the commit in this PR to directly model the structure Daniel  implemented as David and I discusses earlier today.

@dagar , David and I talked through all of your work and you have a an architecture plan that your design fits with, so it makes sense for me to follow your lead in this!  Thanks for the example to follow!

I've tested this work on fmu-v4 and fmu-v5.  We already know that the samv7 code was never completed, so kinetis is the only check left to make if desired.

@davids5, I'll follow up this PR with additional work started in #11260 and #11274 to address the IOCTL issues you have identified.

Let me know what else you need from me in the meantime!

Thanks for all of your help with this.

mcsauder commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1696,7 +1696,7 @@ MavlinkReceiver::handle_message_play_tune(mavlink_message_t *msg)
 
 		if (*tune == 'M') {
-			int fd = px4_open(TONEALARM0_DEVICE_PATH, PX4_F_WRONLY);
+			int fd = px4_open(TONE_ALARM0_DEVICE_PATH, PX4_F_WRONLY);
```
Thanks! I am working to address this in #11274

mcsauder commented on src/systemcmds/tests/test_hrt.cpp:
```diff
@@ -124,10 +124,10 @@ int test_tone(int argc, char *argv[])
 	unsigned long tone;
 
-	fd = px4_open(TONEALARM0_DEVICE_PATH, O_WRONLY);
+	fd = px4_open(TONE_ALARM0_DEVICE_PATH, O_WRONLY);
```
Thanks! I will do more work to address this after #11260 is accepted or closed.

mcsauder commented:
@dagar and @davids5 , rebased against current master (and passing CI!).

mcsauder commented:
Thanks @davids5 !

davids5 commented:
> Thanks @davids5 !

@mcsauder This was super helpful. I did merge on my tablet and missed thanking you. Sorry and Thank you @mcsauder!
